strawberryfields/backends/bosonicbackend/backend.py

```diff
@@ -54,6 +69,205 @@ def __init__(self):
         self._supported["mixed_states"] = True
         self._init_modes = None
         self.circuit = None
+        self.ancillae_samples_dict = {}
+
+    def run_prog(self, prog, **kwargs):
+        """Runs a strawberryfields program using the bosonic backend.
+
+        Args:
+            prog (object): sf.Program instance
+
+        Returns:
+            tuple: list of applied commands,
+                    dictionary of measurement samples,
+                    dictionary of ancilla measurement samples
+
+        Raises:
+            NotApplicableError: if an op in the program does not apply to the bosonic backend
+            NotImplementedError: if an op in the program is not implemented in the bosonic backend
+        """
+        from strawberryfields.ops import (
+            Bosonic,
+            Catstate,
+            DensityMatrix,
+            Fock,
+            GKP,
+            Ket,
+            MbSgate,
+            _New_modes,
+        )
+
+        # Initialize the circuit. This applies all non-Gaussian state-prep
+        self.init_circuit(prog)
+
+        # Apply operations to circuit. For now, copied from LocalEngine;
+        # only change is to ignore preparation classes and ancilla-assisted gates
+        # TODO: Deal with Preparation classes in the middle of a circuit.
+        applied = []
+        samples_dict = {}
+        all_samples = {}
+        for cmd in prog.circuit:
+            nongausspreps = (Bosonic, Catstate, DensityMatrix, Fock, GKP, Ket, _New_modes)
+            ancilla_gates = (MbSgate,)
+            # For ancilla-assisted gates, if they return measurement values, store
+            # them in ancillae_samples_dict
+            if type(cmd.op) in ancilla_gates:
+                # if the op returns a measurement outcome store it in a dictionary
+                val = cmd.op.apply(cmd.reg, self, **kwargs)
+                if val is not None:
+                    for i, r in enumerate(cmd.reg):
+                        if r.ind not in self.ancillae_samples_dict.keys():
+                            self.ancillae_samples_dict[r.ind] = [val[:, i]]
```

Comment:
Suggest checking if using a `defaultdict` could help to decrease the nested nature of this part of the code (that would allow having each member to be initialized as an empty list by default, though suggest checking the quirks of it).

Reply:
Let's keep this on the list of things to investigate in the next round of improvements.